pyspatialopt/models/binary_mclp_distance_matrix.py

```diff
@@ -0,0 +1,128 @@
+    """
+    Generates a dictionary representing the binary coverage of a facility to demand points
+    :param list_dict_facility_demand_distance: (string) A dictionary containing pairwise distance and demand
+    :param dist_threshold：(numeric) The distance threshold
+    :param dl_id_field: (string) The name of the demand point id field in the list_dict_facility_demand_distance object
+    :param fl_id_field: (string) The name of the facility id field in the list_dict_facility_demand_distance object AND fl
+    :param demand_field: (string) The name of demand weight field in the list_dict_facility_demand_distance object
+    :param distance_field: (string) The name of distance in metres field in the list_dict_facility_demand_distance object
+    :param fl_variable_name: (string) The name to use to represent the facility variable
+    :return: (dictionary) A nested dictionary storing the coverage relationships
+    """
+
+    if fl_variable_name is None:
+        fl_variable_name = "facility"
+
+    output = {
+        "version": "1",
+        "type": {
+            "mode": "coverage",
+            "type": "binary",
+        },
+        "demand": {},
+        "totalDemand": 0.0,
+        "totalServiceableDemand": 0.0,
+        "facilities": {fl_variable_name: []}
+    }
+
+    set_facility_id = set()
+    set_demand_id = set()
+    for row in list_dict_facility_demand_distance:
+        set_facility_id.add(str(row[fl_id_field]))
+        # test if this demand id is contained
+        new_demand_id = str(row[dl_id_field])
+        if not new_demand_id in set_demand_id:
+            output["demand"][new_demand_id] = {
+                "area": 0,
+                "demand": float(row[demand_field]),
+                "serviceableDemand": 0.0,
+                "coverage": {fl_variable_name: {}}
+            }
+            set_demand_id.add(new_demand_id)
+        
+    # add facility IDs to facilities
+    for facility_id in set_facility_id:
+        output["facilities"][fl_variable_name].append(facility_id)
+
+    # Determining binary coverage for each demand unit
+    for row in list_dict_facility_demand_distance:
+        if float(row[distance_field]) <= dist_threshold:
+            output["demand"][str(row[dl_id_field])]["serviceableDemand"] = \
+                output["demand"][str(row[dl_id_field])]["demand"]
+            output["demand"][str(row[dl_id_field])]["coverage"][fl_variable_name][str(row[fl_id_field])] = 1
+
+    # summary
+    for row in output["demand"].values():
+        output["totalServiceableDemand"] += row["serviceableDemand"]
+        output["totalDemand"] += row["demand"]    
+    logging.getLogger().info("Binary coverage successfully generated.")
+    return output
+
+def binary_mclp_distance_matrix(file_distance_matrix, service_dist, num_facility, list_field_req = ["facility_id", "demand_id", "demand", "distance"], facility_variable_name = "facility", workspace_path = "."):
```

Comment:
I didn't see this before. This is one of those python gotchas. A list is immutable and should never be used as a default value when it's a parameter. You can read more about it [here](https://docs.python-guide.org/writing/gotchas/)

Instead set `list_field_req=None` and then do a check in the function like:
```python
if list_field_req is None:
    list_field_req = ["facility_id", "demand_id", "demand", "distance"]

Reply:
Good point. Do you mean "a list is mutable and should never be used as a default value"?